src/dpo/src/Optdp.cpp

```diff
@@ -0,0 +1,855 @@
+      int rightPadding = opendp->padRight(inst);
+      arch_->addCellPadding(ndi, leftPadding * siteWidth,
+                            rightPadding * siteWidth);
+    }
+  }
+}
+////////////////////////////////////////////////////////////////
+void Optdp::createLayerMap() {
+  // Relates to pin blockages, etc. Maybe not needed right now.
+  ;
+}
+////////////////////////////////////////////////////////////////
+void Optdp::createNdrMap() {
+  // Relates to pin blockages, etc. Maybe not needed right now.
+  ;
+}
+////////////////////////////////////////////////////////////////
+void Optdp::setupMasterPowers() {
+  // Need to try and figure out which voltages are on the
+  // top and bottom of the masters/insts in order to set
+  // stuff up for proper row alignment of multi-height
+  // cells.  What I do it look at the individual ports
+  // (MTerms) and see which ones correspond to POWER and
+  // GROUND and then figure out which one is on top and
+  // which one is on bottom.  I also record the layers
+  // and use that information later when setting up the
+  // row powers.
+  dbBlock* block = db_->getChip()->getBlock();
+  std::vector<dbMaster*> masters;
+  block->getMasters(masters);
+
+  for (size_t i = 0; i < masters.size(); i++) {
+    dbMaster* master = masters[i];
+
+    double maxPwr = -std::numeric_limits<double>::max();
+    double minPwr = std::numeric_limits<double>::max();
+    double maxGnd = -std::numeric_limits<double>::max();
+    double minGnd = std::numeric_limits<double>::max();
+
+    bool isVdd = false;
+    bool isGnd = false;
+    for (dbMTerm* mterm : master->getMTerms()) {
+      // XXX: Do I need to look at ports, or would the surrounding
+      // box be enough?
+      if (mterm->getSigType() == dbSigType::POWER) {
+        isVdd = true;
+        for (dbMPin* mpin : mterm->getMPins()) {
+          // Geometry or box?
+          double y = 0.5 * (mpin->getBBox().yMin() + mpin->getBBox().yMax());
+          minPwr = std::min(minPwr, y);
+          maxPwr = std::max(maxPwr, y);
+
+          for (dbBox* mbox : mpin->getGeometry()) {
+            dbTechLayer* layer = mbox->getTechLayer();
+            pwrLayers_.insert(layer);
+          }
+        }
+      } else if (mterm->getSigType() == dbSigType::GROUND) {
+        isGnd = true;
+        for (dbMPin* mpin : mterm->getMPins()) {
+          // Geometry or box?
+          double y = 0.5 * (mpin->getBBox().yMin() + mpin->getBBox().yMax());
+          minGnd = std::min(minGnd, y);
+          maxGnd = std::max(maxGnd, y);
+
+          for (dbBox* mbox : mpin->getGeometry()) {
+            dbTechLayer* layer = mbox->getTechLayer();
+            gndLayers_.insert(layer);
+          }
+        }
+      }
+    }
+    int topPwr = RowPower_UNK;
+    int botPwr = RowPower_UNK;
+    if (isVdd && isGnd) {
+      topPwr = (maxPwr > maxGnd) ? RowPower_VDD : RowPower_VSS;
+      botPwr = (minPwr < minGnd) ? RowPower_VDD : RowPower_VSS;
```

Comment:
Its a bit confusing that you are mixing pwr/gnd with vdd/vss.  One set of terminology would be clearer.